### ISSUE TITLE
How to generate genomes.tax for the mOTUs-extender? 

### ISSUE THREAD

**Issue by JSSaini** [motu-tool/mOTUs-extender#2]
Hi Hans, thank you for the mOTUs extender. We would like to add our MAGs to the mOTUs database. I see it requires a specific format file called `genomes.tax`  . Could you please suggest how to generate this file as per the instructions below? 

_A 7 rank taxonomy (NCBI) file which contains the taxonomic assignation of each genome (see here for an example: mOTUs-extender/mOTUs-extender/test/genomes.tax)_

Thank you. 
Jaspreet

**Comment by AlessioMilanese** [comment]
The file should have the following format:
```
genome\tkingdom\tphylum\tclass\torder\tfamily\tgenus\tspecies
```

Where for each taxonomic level you need to have `NCBI taxonomic ID` + space + `NCBI name`.

We used stag for this purpose:
https://github.com/zellerlab/stag

More info: https://github.com/zellerlab/stag/wiki/Classify-genomes

**Comment by JSSaini** [comment]
Hi Alessio, I have obtained the genome annotation file using stag. However, it isn't supported by the mOTUs extender. When running the code, I am getting the following error;

`python ~/git_repositories/mOTUs-extender/mOTUs-extender/motus-extender.py createdb  -w extender -a /media/bioinfoteam/SCRATCH_1TB/jssaini_scratch/github/stag/stag_genomes/fa_files/res_dir/genome_annotation -p LBE_MAGS_DB_14_Feb_2023 -t 16`


```
2023-02-14 13:18:08,864 INFO: Starting mOTUs-extender
2023-02-14 13:18:08,867 ERROR: Taxonomy line for PACBIO_ACC003_d322_2.5K malformed. Should be 7 column rank taxonomy but looks like: No taxonomy found for PACBIO_ACC003_d322_2.5K
2023-02-14 13:18:08,867 INFO: mOTUs-extender finished with error code 1
````

Would you happen to know a workaround? I have also attached the genome annotation file obtained using stag. Thank you so much for your help. 

[genome_annotation.csv](https://github.com/motu-tool/mOTUs-extender/files/10732283/genome_annotation.csv)

**Comment by AlessioMilanese** [comment]
You need to remove the first line (the header), and transform the file from:
```
genome\tkingdom;phylum;class;order;family;genus;species
```
to:
```
genome\tkingdom\tphylum\tclass\torder\tfamily\tgenus\tspecies
```

Also, for the name of the MAGs, I would remove the `./` at the beginning.

Now you will have a file like:
```
d71_METABAT__87-contigs_2.5K.fa	2 Bacteria	1224 Proteobacteria	1236 Gammaproteobacteria	135614 Xanthomonadales	32033 Xanthomonadaceae	83618 Pseudoxanthomonas
d71_METABAT__52-contigs_2.5K.fa	2 Bacteria	976 Bacteroidetes
```

Now it's closer to the required format, but there are only annotations for the known clades. For example for the second MAG there  is the annotation only till phylum level. We can solve this by adding `NA unknown` for the MAGs that do not have an annotation. And we obtain:
```
d71_METABAT__87-contigs_2.5K.fa	2 Bacteria	1224 Proteobacteria	1236 Gammaproteobacteria	135614 Xanthomonadales	32033 Xanthomonadaceae	83618 Pseudoxanthomonas	NA unknown
d71_METABAT__52-contigs_2.5K.fa	2 Bacteria	976 Bacteroidetes	NA unknown	NA unknown	NA unknown	NA unknown	NA unknown
```

**Comment by AlessioMilanese** [comment]
I created the file for you.
[genome_annotation_2.csv](https://github.com/motu-tool/mOTUs-extender/files/10733237/genome_annotation_2.csv)

**Comment by JSSaini** [comment]
Thank you Alessio. I appreciate this.

**Comment by JSSaini** [comment]
Hi Alessio, 
We have provided the input of an external database.  However, we suspect that the program is not detecting our MAGs. Please see the following command we ran, and it also gave us a warning after finishing running. 

Please take a look at the results file I've attached. It doesn't give the reference of the MAGs we added. 

Thank you for your kind assistance.

`motus profile -t 24 -f S1_QC_R1.fastq -r S1_QC_R2.fastq -db /media/bioinfoteam/SCRATCH_850GB/mOTUs3_LBE_DB/extension/extension/LBE_MAGS_DB_2023/db_mOTU/ -p -A -l 50 > d322_cov_A_mOTU3.txt`

```
2023-02-15 15:05:22 Map reads to the marker gene database
   Reads are aligned (by BWA) to marker gene sequences in the reference database
   Reads are filtered for:
    - similarity to the reference gene (>97%), and
    - length of the alignment (at least 50 nucleotides, -l option)

   Number of detected lanes: 1

   2023-02-15 15:05:22 Run bwa on lane 1
     map forward reads: 219.37 s



     map reverse reads: 225.70 s
     sort alignments: 0.10 s
   Total number of reads: 30000878
   Number of reads after filtering: 85085 (0.28 percent)

   Total time for mapping reads: 445.18 s

2023-02-15 15:12:47 Calculate marker gene cluster (MGC) abundance
   Abundance of MGCs are quantified by the number of inserts aligning to their
   member genes
   Inserts can either map to one MGC ('Unique mappers'), or map to many
   MGCs from different species ('Multiple mappers')

   Total number of inserts: 79948
     - Unique mappers: 78790
     - Multiple mappers: 1119
     - Ignored multiple mapper without unique hit: 39

   Total time to calculate the MGCs: 35.41 s

2023-02-15 15:13:22 Generate mOTU profile
   Each mOTU is composed of 6 to 10 MGCs
   The final mOTU insert count is obtained as the median of its MGC read counts

   At least 3 (-g) MGCs need to be detected to consider a mOTU as being present

   Create taxonomy profile
   Warning: mOTU $ not present in the taxonomy

2023-02-15 15:13:30 Finished computation
2023-02-15 15:13:30 Total time: 488.55 s
```
[d322_cov_A_mOTU3.txt](https://github.com/motu-tool/mOTUs-extender/files/10743999/d322_cov_A_mOTU3.txt)

**Comment by AlessioMilanese** [comment]
Could you try to save the mgc reads count (`-M`) and share the file?
Like:
```
motus profile      -M d322_cov_A_mOTU3.MGC     -t 24 -f S1_QC_R1.fastq -r S1_QC_R2.fastq -db /media/bioinfoteam/SCRATCH_850GB/mOTUs3_LBE_DB/extension/extension/LBE_MAGS_DB_2023/db_mOTU/ -p -A -l 50 -o d322_cov_A_mOTU3.txt
```

It seems something went wrong in the creation of the extended database.

**Comment by JSSaini** [comment]
[d322_cov_A_mOTU3.MGC.txt](https://github.com/motu-tool/mOTUs-extender/files/10746354/d322_cov_A_mOTU3.MGC.txt)

Thank you for your response. Please see the attached mgc read count.

**Comment by AlessioMilanese** [comment]
From the file you sent me, I see that there are new MGCs:
```
COG0012.LBE_MAGS_DB_2023_1	95.9193634075
COG0012.LBE_MAGS_DB_2023_10	171.7551395394
COG0012.LBE_MAGS_DB_2023_101	15.0318596441
COG0012.LBE_MAGS_DB_2023_103	143.4859329661
COG0012.LBE_MAGS_DB_2023_105	0.6870205582
[...]
```

It's hard for me to check withous seeing the new database. Is there a way you can share the database in `/media/bioinfoteam/SCRATCH_850GB/mOTUs3_LBE_DB/extension/extension/LBE_MAGS_DB_2023/db_mOTU/`. For example using a ftp server.

**Comment by JSSaini** [comment]
Hi Alessio, you can access the database with the following link:

https://drive.switch.ch/index.php/s/erbcP43PSp5zhKb

**Comment by AlessioMilanese** [comment]
It seems to me that the issue is with the taxonomy of the new mOTUs.
If I check:
```
$ cat db_mOTU_taxonomy_meta-mOTUs.tsv |  grep "LBE_MAGS_DB_2023_61"

LBE_MAGS_DB_2023_61	2 Bacteria	976 Bacteroidetes	NA	NA	NA	NA	NA
```

and the `NA` should be instead `NA unassigned`. 

You can solve with the following command:
```
sed 's/NA\t/NA unknown\t/g' db_mOTU_taxonomy_meta-mOTUs.tsv | sed 's/NA$/NA unknown/'  > temp
mv db_mOTU_taxonomy_meta-mOTUs.tsv db_mOTU_taxonomy_meta-mOTUs.tsv_OLD
mv temp db_mOTU_taxonomy_meta-mOTUs.tsv
```

Now it will be:
```
$ cat db_mOTU_taxonomy_meta-mOTUs.tsv |  grep "LBE_MAGS_DB_2023_61"

LBE_MAGS_DB_2023_61	2 Bacteria	976 Bacteroidetes	NA unknown	NA unknown	NA unknown	NA unknown	NA unknown
```

Nevertheless, the error persist. I will check more next week.

@hjruscheweyh any idea?

**Comment by JSSaini** [comment]
Thank you. Please keep us updated.

**Comment by JSSaini** [comment]
@AlessioMilanese @hjruscheweyh Please let me know if you were able to diagnose the problem. I would appreciate your reply. Thank you.

**Comment by AlessioMilanese** [comment]
The issue is with the file `db_mOTU_MAP_MGCs_to_mOTUs_in-line.tsv` where all the MGCs are under a mOTU `$` (at the end of the file).

Note that the file `db_mOTU_MAP_MGCs_to_mOTUs.tsv` is instead correct.

@hjruscheweyh will have a look in the next few days when he is back from holidays.

**Comment by JSSaini** [comment]
Thank you.

**Comment by hjruscheweyh** [comment]
Closing this issue. Was discussed via mail and fixes were implemented in commit dd15f79bb95a118162a4e247a7c8689bd2f90e93

**Comment by JSSaini** [comment]
Thanks for solving the issue.
